File: JavaAlgorithm/src/main/java/algo/tzashinorpu/FourthRound/Chapter03/threeSum_15.java
package algo.tzashinorpu.FourthRound.Chapter03;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class threeSum_15 {
    public List<List<Integer>> threeSum(int[] nums) {
        // -1,0,1,2,-1,-4 -> -4,-1,-1,0,1,2
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] > 0) break;
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            int l = i + 1;
            int r = nums.length - 1;
            while (l < r) {
                int sum = nums[i] + nums[l] + nums[r];
                if (sum == 0) {
                    res.add(List.of(nums[i], nums[l], nums[r]));
                    while (l < r && nums[l] == nums[l + 1]) l++;
                    while (l < r && nums[r] == nums[r - 1]) r--;
                    l++;
                    r--;
                }
                if (sum > 0) r--;
                if (sum < 0) l++;
            }
        }
        return res;
    }
}

File: Cpp/Algorithm/src/Chapter3/ThreeSum_15.cpp
#include "chapter3.h"
#include <algorithm>
vector<vector<int>> myAlgo::ThreeSum_15::threeSum(vector<int>& nums)
{
	vector<vector<int>> result;
	sort(nums.begin(), nums.end());
	size_t n_size = nums.size();
	for (size_t i = 0; i < n_size; i++)
	{
		if (nums[i] > 0) break;
		if (i > 0 && nums[i] == nums[i - 1]) continue;
		size_t l = i + 1;
		size_t r = n_size - 1;
		while (l<r)
		{
			int sum = nums[i] + nums[l] + nums[r];
			if (sum==0)
			{
				result.push_back({ nums[i], nums[l], nums[r] });
				while (l < r && nums[l] == nums[l + 1]) l++;
				while (l < r && nums[r] == nums[r - 1]) r--;
				l++;
				r--;
			}
			if (sum > 0) r--;
			if (sum < 0) l++;
		}
	}
	return result;
}